python/ray/serve/_private/autoscaling_state.py

```diff
@@ -319,48 +417,126 @@ def register_deployment(
+                        target_num_replicas = get_capacity_adjusted_num_replicas(
+                            deployment_info.deployment_config.num_replicas,
+                            deployment_info.target_capacity,
+                        )
+                        deployment_autoscaling_state.register(
+                            deployment_info,
+                            target_num_replicas,
+                            is_part_of_autoscaling_application=True,
+                        )
+                    self._deployment_autoscaling_states[
+                        deployment_id
+                    ] = deployment_autoscaling_state
+
+                deployment_autoscaling_states[
+                    deployment_id
+                ] = self._deployment_autoscaling_states[deployment_id]
+
+            self._app_autoscaling_states[app_name] = ApplicationAutoscalingState(
+                app_name, deployment_autoscaling_states
+            )
+
+        self._app_autoscaling_states[app_name].register(config)
+
+    def deregister_application(
+        self, app_name: ApplicationName, deployment_names: List[str]
+    ):
+        """Remove application from tracking."""
+        for deployment_name in deployment_names:
+            deployment_id = DeploymentID(deployment_name, app_name)
+            self._deployment_autoscaling_states.pop(deployment_id, None)
+        self._app_autoscaling_states.pop(app_name, None)
+
+    def get_scaling_decisions_for_application(
+        self, app_name: ApplicationName, deployments: Dict[str, DeploymentDetails]
+    ) -> Dict[str, int]:
+        return self._app_autoscaling_states[app_name].get_scaling_decisions(deployments)
+
+    def is_part_of_autoscaling_application(self, deployment_id: DeploymentID):
+        return deployment_id.app_name in self._app_autoscaling_states
 
     def update_running_replica_ids(
         self, deployment_id: DeploymentID, running_replicas: List[ReplicaID]
     ):
-        self._autoscaling_states[deployment_id].update_running_replica_ids(
+        self._deployment_autoscaling_states[deployment_id].update_running_replica_ids(
             running_replicas
         )
 
     def on_replica_stopped(self, replica_id: ReplicaID):
         deployment_id = replica_id.deployment_id
-        if deployment_id in self._autoscaling_states:
-            self._autoscaling_states[deployment_id].on_replica_stopped(replica_id)
+        if deployment_id in self._deployment_autoscaling_states:
+            self._deployment_autoscaling_states[deployment_id].on_replica_stopped(
+                replica_id
+            )
 
     def get_metrics(self) -> Dict[DeploymentID, float]:
         return {
             deployment_id: self.get_total_num_requests(deployment_id)
-            for deployment_id in self._autoscaling_states
+            for deployment_id in self._deployment_autoscaling_states
         }
 
-    def get_target_num_replicas(
+    def get_scaling_decision_for_deployment(
```

Comment:
curious why we changed the function name here? I thought `get_target_num_replicas` is more appropriate

Reply:
we also have get_scaling_decision_for_application in the same ASM class so renamed to get_scaling_decision_for_deployment for consistency.